_release/bat/base_bat/base_bat_test.go

```diff
@@ -311,6 +327,185 @@ func TestDeleteAllInstances(t *testing.T) {
+	// support for meta data improves.
+	options := bat.ImageOptions{
+		Name: name,
+	}
+	//downaload cirros image
+	ImageFile, err := os.Create("cirros-0.3.4-x86_64-disk.img")
+	if err != nil {
+		t.Fatalf("Error while creating image file %v", err)
+	}
+	defer ImageFile.Close()
+
+	resp, err := http.Get("http://download.cirros-cloud.net/0.3.4/cirros-0.3.4-x86_64-disk.img")
+	if err != nil {
+		t.Fatalf("Error while downloading image file %v", err)
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(ImageFile, resp.Body)
+
+	if err != nil {
+		t.Fatalf("Error while saving image file %v", err)
+	}
+
+	img, err := bat.AddImage(ctx, "", "cirros-0.3.4-x86_64-disk.img", &options)
+	if err != nil {
+		t.Fatalf("Unable to add image %v", err)
+	}
+
+	defer deleteCustomData(ctx, img.ID)
+
+	if img.ID == "" || img.Name != name || img.Status != "active" ||
+		img.Visibility != "public" || img.Protected {
+		t.Errorf("Meta data of added image is incorrect")
+	}
+	//get data to create workload yaml file
+	defaults := bat.DefaultResources{
+		VCPUs: 2,
+		MemMB: 128,
+	}
+
+	source := bat.Source{
+		Type: "image",
+		ID:   img.ID,
+	}
+	disk := bat.Disk{
+		Bootable:  true,
+		Source:    &source,
+		Ephemeral: true,
+	}
+
+	WLopts := bat.WorkloadOptions{
+		Description:     "BAT VM Test",
+		VMType:          "qemu",
+		FWType:          "legacy",
+		Defaults:        defaults,
+		CloudConfigFile: "CirrosInit.yaml",
+		Disks:           []bat.Disk{disk},
+	}
+	y, err := yaml.Marshal(WLopts)
+	if err != nil {
+		t.Fatal("fail to read yaml file $v", err)
+	}
+	ioutil.WriteFile("CirrosWorkload.yaml", y, 0644)
+	ioutil.WriteFile("CirrosInit.yaml", []byte(vmCloudInit), 0644)
+
+	//create the workload
+	args := []string{"workload", "create", "-yaml", "CirrosWorkload.yaml"}
+	_, err = bat.RunCIAOCLIAsAdmin(ctx, "", args)
+	if err != nil {
+		t.Fatal("Fail to create workload $v", err)
+
+	}
+	//Get ID of workload just created
+	CurWL, err := bat.GetWorkloadByName(ctx, "", WLopts.Description)
+	if err != nil {
+		t.Fatalf("Can't find data for workload %v", err)
+	}
+	//Launch instance of that workload
+	instance, err := bat.LaunchInstances(ctx, "", CurWL.ID, 1)
+	if err != nil {
+		t.Fatal("Fail to launch instance: $v", err)
+	}
+	_, err = bat.WaitForInstancesLaunch(ctx, "", instance, false)
+	if err != nil {
+		t.Errorf("Instance %s did not launch: %v", instance[0], err)
+	}
+	//Access the instances with the user and pasword specified in the could init file
+	config := &ssh.ClientConfig{
+		User: "ciaouser",
+		Auth: []ssh.AuthMethod{
+			ssh.Password("abc123"),
+		},
+	}
+	Template := `{{$x := filterContains . "Status" "active"}}{{range $x}}{{.SSHIP}}:{{.SSHPort}}{{end}}`
+	argsInstance := []string{"instance", "list", "-f", Template}
+	InstanceIP, err := bat.RunCIAOCLIAsAdmin(ctx, "", argsInstance)
+	if err != nil {
+		t.Fatal("Fail to create get Instance's data $v", err)
+	}
+
+	client, err := ssh.Dial("tcp", string(InstanceIP[:]), config)
```

Comment:
Could we move all this ssh stuff into a separate function and place it in the BAT package.  This function will be generically useful for lots of different tests.  You're going to need to split up this function anyway to get it past gocyclo.